src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,672 @@
+	}
+	m_cts = state;
+}
+
+DECLARE_WRITE_LINE_MEMBER( z180asci_channel_base::dcd_wr )
+{
+	if (m_id)
+		return;
+
+	// In extended mode, DCD autoenables RX if configured
+	if (m_ext && (m_asci_ext && Z180_ASEXT_DCD0) == 0)
+	{
+		m_rx_enabled = state ? false : true;
+		if (state)
+			m_asci_ext &= ~(Z180_ASEXT_BRK_DET);
+	}
+
+	m_dcd = state;
+	if (m_dcd)
+		m_irq = 1;
+}
+
+DECLARE_WRITE_LINE_MEMBER( z180asci_channel_base::rxa_wr )
+{
+	m_rxa = state;
+}
+
+DECLARE_WRITE_LINE_MEMBER( z180asci_channel_base::brg_wr )
+{
+	m_cka_handler(state);
+	if (!state)
+		transmit_edge();
+	else
+		receive_edge();
+}
+
+
+DECLARE_WRITE_LINE_MEMBER( z180asci_channel_base::cka_wr )
+{
+	// For channel 1, CKA can be disabled
+	if (m_id && (m_asci_cntla && Z180_CNTLA1_CKA1D)) return;
+
+	if(state != m_clock_state) {
+		m_clock_state = state;
+		if(!state)
+			transmit_edge();
+		else
+			receive_edge();
+	}
+}
+
+void z180asci_channel_base::transmit_edge()
+{
+	m_tx_counter++;
+	if (m_tx_counter != m_divisor) return;
+	m_tx_counter = 0;
+
+	if (m_asci_cntla & Z180_CNTLA_TE)
+	{
+		switch (m_tx_state)
+		{
+		case STATE_START:
+			output_txa(0);
+			m_tx_state = STATE_DATA;
+			m_tx_parity = 0;
+			break;
+		case STATE_DATA:
+			m_tx_bits++;
+			output_txa(BIT(m_tsr, 0));
+			m_tx_parity ^= BIT(m_tsr, 0);
+			m_tsr >>= 1;
+			if (m_tx_bits == ((m_asci_cntla & Z180_CNTLA_MODE_DATA) ? 8 : 7))
+			{
+				m_tx_state = (m_asci_cntlb & Z180_CNTLB_MP) ? STATE_MPB : (m_asci_cntla & Z180_CNTLA_MODE_PARITY) ? STATE_PARITY : STATE_STOP;
+				m_tx_bits = (m_asci_cntla & Z180_CNTLA_MODE_STOPB) ? 2 : 1;
+			}
+			break;
+		case Z180_CNTLB_MP:
+			m_tx_state = STATE_STOP;
+			output_txa(m_asci_cntlb & Z180_CNTLB_MPBT ? 1 : 0);
+			break;
+		case STATE_PARITY:
+			m_tx_state = STATE_STOP;
+			if (m_asci_cntlb & Z180_CNTLB_PEO) m_tx_parity ^= 1; // odd parity
+			output_txa(m_tx_parity);
+			break;
+		case STATE_STOP:
+			if (m_tx_bits)
+			{
+				m_tx_bits--;
+				if (m_tx_bits == 0)
+				{
+					output_txa(1);
+					m_asci_stat |= Z180_STAT_TDRE;
+					if (m_asci_stat & Z180_STAT_TIE)
+					{
+						m_irq = 1;
+					}
+				}
+			}
+			break;
+		case STATE_BREAK:
+			output_txa(0);
+			break;
+		}
+	}
+}
+
+void z180asci_channel_base::receive_edge()
+{
+	m_rx_counter++;
+	if (m_rx_counter != m_divisor) return;
+	m_rx_counter = 0;
+
+	if (!m_rx_enabled) return;
+
+	if (m_asci_cntla & Z180_CNTLA_RE)
+	{
+		switch (m_rx_state)
+		{
+		case STATE_START:
+			if(m_rxa == 0)
+			{
+				m_rx_state = STATE_DATA;
+				m_rx_bits = 0;
+				m_rx_parity = 0;
+				m_rx_error = 0;
+				m_rsr = 0;
+			}
+			break;
+		case STATE_DATA:
+			m_rsr |= m_rxa << m_rx_bits;
+			m_rx_parity ^= m_rxa;
+			m_rx_bits++;
+			if (m_rx_bits == ((m_asci_cntla & Z180_CNTLA_MODE_DATA) ? 8 : 7))
+			{
+				m_rx_bits = (m_asci_cntla & Z180_CNTLA_MODE_STOPB) ? 2 : 1;
+				m_rx_state = (m_asci_cntlb & Z180_CNTLB_MP) ? STATE_MPB : (m_asci_cntla & Z180_CNTLA_MODE_PARITY) ? STATE_PARITY : STATE_STOP;
+			}
+			break;
```

Comment:
This will break if the format is changed from 8 to 7 at the wrong moment.  It will compare to 7, see that it doesn’t have a complete byte, then when it gets back here on the next bit cell (after the data format has been changed to 7-bit), and then compare to 8 and still think it doesn’t have a complete byte.  It will then be stuck in this state.  That isn’t robust.